### DISCUSSION TITLE
Specific API error response codes from Mailu admin container instead of generic 500

### DISCUSSION THREAD

**Discussion by StuHoward** [Mailu/Mailu#3668] [User support] [answered]
Is there a way to pass a more specific response code back from the Mailu 'admin' container when an API request can't be fulfilled?  At the moment I am getting a simple 500 'Internal Server Error' for a variety of scenarios resulting from API calls.  For example, when trying to delete a malformed alias 'badalias' the admin container produces the following:

```
[2024-11-20 10:51:48,713] ERROR in app: Exception on /api/v1/alias/badalias [DELETE]
...
  File "/app/mailu/models.py", line 63, in process_bind_param
    raise ValueError('invalid email address (no "@")')
ValueError: invalid email address (no "@")
```
...but the NGINX 'front' container only gets a generic 500 error:

`10.x.x.x - - [20/Nov/2024:10:51:48 +0000] "DELETE /api/v1/alias/badalias HTTP/1.1" 500 37 "-" "GuzzleHttp/7"`

I would like this to return a HTTP status 400 "Bad Request" error so that the calling code can take appropriate action based on the response.

**Another example:**

When trying to create a user which already exists, the admin container produces the following:

```
[2024-11-20 10:51:55,364] ERROR in app: Exception on /api/v1/user [POST]
Traceback (most recent call last):
  File "/app/venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/app/venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "user_pkey"
DETAIL:  Key (email)=(01926719-0925-72c5-bf92-e65887f1c3b0@mydomain.com) already exists.
```
...and the front container again receives only a generic 500 error:

`10.x.x.x - - [20/Nov/2024:10:51:55 +0000] "POST /api/v1/user HTTP/1.1" 500 37 "-" "GuzzleHttp/7"`

In this case I would like it to return a HTTP 409 "Conflict" response, so that the calling code can take a different course of action.

Is there a way to achieve this?

**Comment by Diman0** [comment]
Are you using the latest version of Mailu?
The API has many validation checks. Both examples should already be part of the validation checks:

Invalid destination email address for alias:
https://github.com/Mailu/Mailu/blob/92b6a050fc706e619121254deb69854ff8a5334b/core/admin/mailu/api/v1/alias.py#L137

User already exists:
https://github.com/Mailu/Mailu/blob/92b6a050fc706e619121254deb69854ff8a5334b/core/admin/mailu/api/v1/user.py#L123

**Comment by Diman0** [comment, marked as answer]
I cannot replicate your examples via the Mailu demo server: https://test.mailu.io/api/v1/
See the URL https://mailu.io/2024.06/demo.html for the API  token.


```curl -X 'DELETE' \
  'https://test.mailu.io/api/v1/alias/badalias' \
  -H 'accept: application/json' \
  -H 'Authorization: APITokenForMailu'
```
results in the response
```
{
  "code": 400,
  "message": "Provided alias (email address) badalias is not a valid email address"
}
```

The previous Mailu 2.0 version did not have all the validation checks yet. These were introduced in Mailu 2024.06

**Comment by StuHoward** [comment]
Ah, I'm using version 2.0 so I guess the answer is that I need to upgrade.

Many thanks for your help.
